crates/consensus/beacon/src/engine/prune.rs

```diff
@@ -0,0 +1,160 @@
+//! Prune management for the engine implementation.
+
+use futures::{FutureExt, Stream};
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use reth_prune::{Pruner, PrunerError, PrunerWithResult};
+use reth_tasks::TaskSpawner;
+use std::task::{ready, Context, Poll};
+use tokio::sync::oneshot;
+use tracing::trace;
+
+/// Manages pruning under the control of the engine.
+///
+/// This type controls the [Pruner].
+pub(crate) struct EnginePruneController<St> {
+    /// The current state of the pruner.
+    pruner_state: PrunerState<St>,
```

Comment:
why can't the pruner just live on its own task?

because we need tight control over when it's running and when not?

Reply:
yes, correct. Otherwise we won't be able to control who owns the write tx, OR we will have a more complicated flow like that:
1. Pruner is notified with `CanonStateNotification` by Blockchain Tree
2. Pruner notifies Sync Controller that it wants to start the pruning process with `RequestPruneStart`
3. Sync Controller processes the `RequestPruneStart` message, sets  `self.is_pruning = true` disallowing the Engine to process any incoming FCUs, replies with `ResponsePruneStart`
4. Pruner starts the pruning process and sends `PruneFinished` when finishes
5. Sync Controller processes the `PruneFinished` message, sets `self.is_Pruning = false` allowing the Engine to continue processing incoming FCUs

Reply:
> yes, correct. Otherwise we won't be able to control who owns the write tx

I see, yeah that makes sense

Reply:
> Pruner is notified with CanonStateNotification by Blockchain Tree

hmm, if the beacon engine controls this, then it doesn't need it? I think that's my main confusion, why does it even need this when the pruner is under the control of the engine and it only needs to know the local head

Reply:
> hmm, if the beacon engine controls this, then it doesn't need it?

ah, yes, that totally makes sense. We can just pass the local head from the engine.
But I would still prefer to leave the `Pruner::check_tip` method, so we would have the logic of minimum pruning interval inside `Pruner`. It will just accept the `tip_block_number` from the caller now.

Reply:
so we have

`BeaconEngine` polls `PruneController` with head number, `PrunerController` can spawn a task to prune depending on the head number
`BeaconEngine` needs to prevent commits while pruner is active

Reply:
correct. So there needs to be a way to communicate to `BeaconEngine` that the pruner is currently running, and I chose the same way as with pipeline: have a separate controller which spawns the pruner and contains the current state (idle/running).